File: app/assets/javascripts/models/tag.js
App.Tag = Ember.Object.extend({
  index: null,
  project: null,

  id: function() {
    var index = this.get('index');
    if (!Ember.isNone(index)) {
      return index + 1;
    }
  }.property('index'),

  name: function(key, value) {
    var index = this.get('index'),
        tags = this.get('project.tags');

    if (arguments.length === 2) {
      tags = tags.slice();
      tags[index] = value;
      this.set('project.tags', tags);
    } else {
      value = tags[index];
    }

    return value;
  }.property('index', 'project.tags')
});

File: app/models/tag.rb
class Tag
  include ActiveModel::SerializerSupport
  
  def initialize tags, index
    @tags, @index = tags, index
  end

  def id
    @index + 1
  end

  def project
    @tags.project
  end

  def name
    @tags.values[@index]
  end

  def name= value
    @tags.values[@index] = value
  end

  def eql? other
    other.class == self.class && other.project == project && other.id == id
  end

  alias_method :==, :eql?
  alias_method :equal?, :eql?
  alias_method :===, :eql?
end
